### ISSUE TITLE
Add deployment guide for Prefect 2.0

### ISSUE THREAD

**Issue by deepyaman** [kedro-org/kedro#2431]
## Description
The existing deployment guides cover Prefect deployment, but not Prefect 2.0, and there have been a number of questions recently about this (e.g. https://www.linen.dev/s/kedro/t/9703126/hi-everyone-does-anyone-know-whether-kedro-supports-prefect-#d8f1e73e-ee9f-4399-ae4b-31f75d657879).

## Context
Prefect 2.0 deployment should look different from deploying to the Prefect 1.x series, so it's not as simple as slightly modifying the deployment instructions. Furthermore, the Prefect deployment guide creates underlying Prefect objects (i.e. not how it's demonstrated in the tutorial, so it requires looking at Prefect source code to update properly).

## Possible Implementation
N/A

## Possible Alternatives
Support Prefect 2.0 as a built-in runner.

**Comment by astrojuanlu** [comment]
Related: do we want to keep the existing Prefect 1.0 guide? I'm not familiar with the project enough to judge.

**Comment by deepyaman** [comment]
> Related: do we want to keep the existing Prefect 1.0 guide? I'm not familiar with the project enough to judge.

I'm not familiar enough, either, but I assume most new deployments should be to Prefect 2.0 (so I'd say it's not as important to keep Prefect 1.0).

**Comment by ofir-insait** [comment]
What about simply calling `os.system("kedro run")` from a Prefect workflow? What are the disadvantages of doing such thing?

**Comment by astrojuanlu** [comment]
@ofir-insait from my limited understanding of Prefect, I don't think there would be anything wrong with doing it, just a loss of granularity compared to with, say, a Kedro pipeline transformed into a Prefect flow.

**Comment by ofir-insait** [comment]
Gotcha, thanks!

**Comment by jmalovera10** [comment]
Hi! I wanted to ask if the Prefect deployment guide is going to be updated soon? given that Prefect Cloud 1.0 is going to be frozen on May 15th 2023.

**Comment by hugocool** [comment]
One point to consider is that Prefect 2.0 server is not open source, so one has to use their cloud.
Therefore, for sensitive applications (like healthcare) prefect 2.0 is not an option, and prefect 1.x is actually preferred.
And while their prefect 1.0 server may be sunset, the version you deploy yourself is still a great option all things considered.

**Comment by astrojuanlu** [comment]
@hugocool thanks for chiming in (see [the Slack conversation](https://www.linen.dev/s/kedro/t/12115934/there-is-a-bug-in-the-kedro-prefect-deployment-when-using-ke#8935f1b0-0b01-4c1f-8224-26a07e335319) for more context). I went and checked and apparently there _is_ a Prefect 2.0 server:

https://github.com/PrefectHQ/server, https://docs.prefect.io/2.10.11/host/

Or am I missing something?

About Prefect 1.0 transition: https://www.prefect.io/guide/blog/freezing-legacy-prefect-cloud-1-accounts-on-starter-and-standard-plans/

> Today we are announcing that legacy Prefect Cloud 1 accounts will be frozen for some users starting May 15th 2023.

@jmalovera10 we have limited capacity and unfortunately updating the Prefect docs is not our priority in the short term, but we would be more than happy to review community contributions.

**Comment by hugocool** [comment]
> @hugocool thanks for chiming in (see [the Slack conversation](https://www.linen.dev/s/kedro/t/12115934/there-is-a-bug-in-the-kedro-prefect-deployment-when-using-ke#8935f1b0-0b01-4c1f-8224-26a07e335319) for more context). I went and checked and apparently there _is_ a Prefect 2.0 server:
> 
> https://github.com/PrefectHQ/server, https://docs.prefect.io/2.10.11/host/
> 
> Or am I missing something?
> 
> About Prefect 1.0 transition: https://www.prefect.io/guide/blog/freezing-legacy-prefect-cloud-1-accounts-on-starter-and-standard-plans/
> 
> > Today we are announcing that legacy Prefect Cloud 1 accounts will be frozen for some users starting May 15th 2023.
> 
> @jmalovera10 we have limited capacity and unfortunately updating the Prefect docs is not our priority in the short term, but we would be more than happy to review community contributions.

The repo you mentioned is for prefect 1.0, if you look at  `https://github.com/PrefectHQ/server/blob/master/setup.py`, you'll see 

```
install_requires = [
    "prefect >= 1.0.0, < 2.0.0",
```
Unfortunately prefect server 2.0 is not open source.

**Comment by astrojuanlu** [comment]
I asked for clarification of the Prefect team and they reassured that `$ prefect server` is the open source Prefect 2.0 server, and are clarifying the docs https://github.com/PrefectHQ/server/pull/418

I'm hiding these comments as they are off-topic, no offence 🙏🏽

**Comment by noklam** [comment]
Adding this for reference
![image](https://github.com/kedro-org/kedro/assets/18221871/65bebd09-8b2a-4d55-ad3e-2bcbdaa928d8)

From this [Slack thread](https://kedro-org.slack.com/archives/C03RKP2LW64/p1687005089782929?thread_ts=1686997990.822149&cid=C03RKP2LW64)

**Comment by jmalovera10** [comment]
> @hugocool thanks for chiming in (see [the Slack conversation](https://www.linen.dev/s/kedro/t/12115934/there-is-a-bug-in-the-kedro-prefect-deployment-when-using-ke#8935f1b0-0b01-4c1f-8224-26a07e335319) for more context). I went and checked and apparently there _is_ a Prefect 2.0 server:
> 
> https://docs.prefect.io/2.10.11/host/
> 
> About Prefect 1.0 transition: https://www.prefect.io/guide/blog/freezing-legacy-prefect-cloud-1-accounts-on-starter-and-standard-plans/
> 
> > Today we are announcing that legacy Prefect Cloud 1 accounts will be frozen for some users starting May 15th 2023.
> 
> @jmalovera10 we have limited capacity and unfortunately updating the Prefect docs is not our priority in the short term, but we would be more than happy to review community contributions.

@astrojuanlu thank you for replying! I understand if it is not in the priorities right now. Maybe I could help, I managed to make working example for the project I am working on. Should I make a PR for documentation?

**Comment by astrojuanlu** [comment]
> Should I make a PR for documentation?

@jmalovera10 Yes please! 💖

**Comment by astrojuanlu** [comment]
Reviews welcome on gh-2725 🚀

**Comment by hugocool** [comment]
@jmalovera10 , thanks for including me/us.
We just finished two deployments of kedro; on AWS batch, and Prefect v1.
From my perspective, the deployment on prefect is not feature complete, specifically with respect to [filtering](https://docs.kedro.org/en/stable/kedro.pipeline.Pipeline.html#kedro.pipeline.Pipeline.filter). Because for as far as I know one of the main differences between the batch and prefect deployments is that batch runs every node as a single job, which is not ideal but does keep the filtering feature working, but for prefect it runs a pipeline as a flow so it wasn't clear to me whether it could run single nodes and/or entire pipelines allowing one to keep using the traditional kedro run commands. We [also found some issues with node hooks, specifically when using kedro-mlflow](https://kedro-org.slack.com/archives/C03RKPCLYGY/p1685527907180559).
basically we would like to contribute a plugin for this deployment.

also, did you figure out the mapping of the different agents used in v1 vs v2? so the dockeragent, localagent, kubernetesagent?

**Comment by astrojuanlu** [comment]
@hugocool your feedback is really appreciated and I'm hoping @jmalovera10 can comment on that. In the meantime, we have merged gh-2748 which explains a simple approach that works on Prefect 2.0, hence closing the original scope of this issue. Feel free to keep discussing, or otherwise open a new issue for further improvements.

**Comment by jmalovera10** [comment]
> @jmalovera10 , thanks for including me/us. We just finished two deployments of kedro; on AWS batch, and Prefect v1. From my perspective, the deployment on prefect is not feature complete, specifically with respect to [filtering](https://docs.kedro.org/en/stable/kedro.pipeline.Pipeline.html#kedro.pipeline.Pipeline.filter). Because for as far as I know one of the main differences between the batch and prefect deployments is that batch runs every node as a single job, which is not ideal but does keep the filtering feature working, but for prefect it runs a pipeline as a flow so it wasn't clear to me whether it could run single nodes and/or entire pipelines allowing one to keep using the traditional kedro run commands. We [also found some issues with node hooks, specifically when using kedro-mlflow](https://kedro-org.slack.com/archives/C03RKPCLYGY/p1685527907180559). basically we would like to contribute a plugin for this deployment.
> 
> also, did you figure out the mapping of the different agents used in v1 vs v2? so the dockeragent, localagent, kubernetesagent?

@hugocool thanks for the question. The [deployment script I adapted from Prefect 1.0](https://docs.kedro.org/en/stable/deployment/prefect.html#convert-your-kedro-pipeline-to-prefect-2-0-flow) takes a pipeline and wraps Kedro nodes as Prefect tasks. This helps to keep consistent mappings between pipelines -> flows and nodes -> tasks. In the filtering example you propose, it could be done by initializing the filtered pipeline and [extracting the topological order of the pipeline](https://github.com/kedro-org/kedro/blob/bbd77dbc2d8d04938e3e2b987fc6281e4d07a244/kedro/pipeline/pipeline.py#L355). Then you could wrap the nodes as tasks and execute them by topological "layers", in other words, execute the nodes that are the dependencies of other nodes. If you were to modify the [script](https://docs.kedro.org/en/stable/deployment/prefect.html#convert-your-kedro-pipeline-to-prefect-2-0-flow) to achieve this, you could replace these pipeline declarations:

```python
pipeline = pipelines.get(pipeline_name)
```
For this one:

```python
pipeline = filter(...)
```
The limitation of this approach is that you must have a pipeline for the script to work. However, the filtering functions allow you to filter by node names, so providing a single node name would suffice for single node executions.

On the other hand, I have only worked with the [Process Block](https://docs.prefect.io/2.10.18/concepts/infrastructure/#process) that executes a local entrypoint given an execution event received by an [Agent](https://docs.prefect.io/2.10.18/concepts/work-pools/#agent-overview). What changed in Prefect 2.0 is that the Agent is independent from the infrastructure, so [you rely on a single Agent type and define the entrypoint of the infrastructure you want to use](https://docs.prefect.io/2.10.18/guides/migration-guide/#what-changed-in-flow-deployment-patterns). I haven't tried, but my guess is that by configuring, for example, a [Docker Block](https://docs.prefect.io/2.10.18/guides/deployment/docker/?h=docker) the deployment should be similar as for a Process Block. Let me know if this is helpful 😄

**Comment by astrojuanlu** [comment]
Heads up for people subscribed to this issue, there seem to be some issues with Kedro on Prefect under Windows https://github.com/kedro-org/kedro/issues/2346
